### DISCUSSION TITLE
30.2 Mutable static variables

### DISCUSSION THREAD

**Discussion by njr0** [google/comprehensive-rust#1297] [General]
Just staring at the code, it isn't really clear what the potential data races are here. After all, this is a sequential program that first increments something, then prints it.

I think you either need to explain what the potential races are, or suggest people take out the `unsafe` keywords and see what the compiler says:

```
   Compiling playground v0.0.1 (/playground)
error[E0133]: use of mutable static is unsafe and requires unsafe function or block
 --> src/main.rs:4:8
  |
4 |      { COUNTER += inc; }  // Potential data race!
  |        ^^^^^^^^^^^^^^ use of mutable static
  |
  = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior

error[E0133]: use of mutable static is unsafe and requires unsafe function or block
  --> src/main.rs:10:27
   |
10 |      { println!("COUNTER: {COUNTER}"); }  // Potential data race!
   |                           ^^^^^^^^^ use of mutable static
   |
   = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior
   = note: this error originates in the macro `$crate::format_args_nl` which comes from the expansion of the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0133`.
error: could not compile `playground` (bin "playground") due to 2 previous errors
```

... which is basically that it would be a problem if the program were multi-threaded (which it isn't!)

**Comment by mgeisler** [comment]
> Just staring at the code, it isn't really clear what the potential data races are here. After all, this is a sequential program that first increments something, then prints it.

True! The point here is that Rust does all its analysis in isolation. So even though this example is safe, the compiler doesn't know this — it basically assumes the worst case, which is that some other thread will mutate the variable concurrently.
 
> I think you either need to explain what the potential races are, or suggest people take out the `unsafe` keywords and see what the compiler says:

Taking out the `unsafe` is a good idea, thanks. I created #1331 to mention this.
